### PR TITLE
Support for Kernel build with vexriscv-smp.

### PR DESCRIPTION
**Pull request by g2gps**
## Summary

Adds support for building the Litex port with the vexriscv-smp softcore. In short, this entails:
 - Support for building kernel, and user space applications.
 - Enables MMU.
 - vexriscv_smp core runs in supervisor mode.
 - Applications are built using the `nuttx export` target. These are currently ran from a ramdisk, mounted on startup.
 - Enables Nuttx to be run as a payload for OpenSBI.
 - Adds support for PLIC interrupt controller.
 - Separate kernel configuration.
 - New documentation for both cores.

## Impact

Changes to the Litex build with the current vexriscv core:
 - Scheduler timer is now controlled with the standard `timer0` built-in by the default Litex configuration for vexriscv. 
 - No other changes. The core type is selected by a Kconfig option.

## Testing

`vexriscv` and `vexriscv_smp` cores both running on arty_a7. 

From what I've tested, the port seems to work fine. We'll be using this core intensely in the near future. I'm happy to push any bug fixes or improvements we find as time goes on.

### CODE REVIEW

File: boards/risc-v/qemu-rv/rv-virt/README.txt [hunk 1]
```diff
@@ -20,14 +20,17 @@
...
   $ cd nuttx
   $ make distclean
-  $ ./tools/configure.sh rv-virt:knsh64
+  $ # For 64-bit build.
+  $ ./tools/configure.sh rv-virt:knsh64 
+  $ # For 32-bit build.
+  $ ./tools/configure.sh rv-virt:knsh32
   $ make V=1 -j7
   $ make export V=1
   $ cd ../apps
```
Comment (acassis) [inline boards/risc-v/qemu-rv/rv-virt/README.txt:36]: Please consider moving it to Documentation/ to be displayed at https://nuttx.apache.org/docs/latest/platforms/risc-v/index.html
We need to phase out the README.txt from boards/

**Comment by masayuki2009** [comment]
>qemu for whatever reason allows/handles the access from S -> M mode, maybe @masayuki2009 can explain how that >actually works. I think there is something in the command line options that allows this.

In qemu-rv implementation with BUILD_KERNEL, we call `up_mtimer_initialize()` before transiting to S-mode.
In addition, I added `qemu_rv_exception_m.S` to handle `qemu_rv_mtimer_interrupt()` which sends an S-mode software interrupt. In `qemu_rv_ssoft_interupt(), it calls `nxsched_process_timer()` if g_stimer_pending is true.

As you can see in the `qemu_rv_ssoft_interupt()`, it also needs to handle IPI (Inter-Processor Interrupt) for SMP which we assume to use ACLINT instead of CLINT. Because ACLINT can send IPI in S-mode directly.

https://github.com/riscv/riscv-aclint/blob/main/riscv-aclint.adoc

**Comment by AndrewD** [comment]
> In qemu-rv implementation with BUILD_KERNEL, we call `up_mtimer_initialize()` before transiting to S-mode. In addition, I added `qemu_rv_exception_m.S` to handle `qemu_rv_mtimer_interrupt()` which sends an S-mode software interrupt. In `qemu_rv_ssoft_interupt(), it calls `nxsched_process_timer()` if g_stimer_pending is true.
> 
> As you can see in the `qemu_rv_ssoft_interupt()`, it also needs to handle IPI (Inter-Processor Interrupt) for SMP which we assume to use ACLINT instead of CLINT. Because ACLINT can send IPI in S-mode directly.
> 
> https://github.com/riscv/riscv-aclint/blob/main/riscv-aclint.adoc

Thanks for the detail - this explains the differences we've been observing between Vexriscv and qemu. We are still learning how this works and appreciate the information.

**Comment by pussuw** [comment]
> > stripped down SBI for NuttX
> 
> This sounds very interesting: integrating with the build system is nice and clean and interesting for us. If you are open to sharing this we are definitely interested. Ultimately our focus is just to get this working, so if we can leverage your experience and existing work that would be greatly appreciated.

Sure I can provide what I have. I use the stripped version for debugging myself so it has been "somewhat" maintained. There are some issues that need to be fixed though, especially support for SMP mode needs patching. I will not be implementing any of this myself so if you decide to use it it is up to you to maintain.

Keep in mind, that it is a VERY STRIPPED down version. The SBI host has many responsibilities like resolving and emulating instructions that are not implemented into the CPU itself, and this stripped version does none of that. I still _STRONGLY_ suggest using OpenSBI (or something similar) as it is a complete implementation, unless you are 100% sure you wont need its wide feature set.

**Comment by g2gps** [comment]
@pussuw Thanks for the feedback and your port on nuttSBI. Admittedly, I had underestimated exactly what was required to get things working correctly.

I ended up finding and using the [Litex port of openSBI](https://github.com/litex-hub/opensbi/tree/0.8-linux-on-litex-vexriscv), which is the same used in the [Linux on Litex Vexriscv](https://github.com/litex-hub/linux-on-litex-vexriscv) project. 

Currently, the port works, however I still need to integrate openSBI into the build system. There's a few ways I could do this:

1. Add it into the current `Make.defs` under `arch/risc-v/src/opensbi`. Keeping in mind that it would be pointing to a different repository.
2. Add something similar as (1) in `arch/risc-v/src/litex`.
3. Add instructions on where it it, and how to build it. Let the end-user go through this process.

The openSBI binary is currently loaded to gateware separately from the kernel, so I'm in favour of (3). However, I've can understand the desire to manage the build process as part of Nuttx. In that case, should the built openSBI binary just be placed in the root `nuttx` directory, along with the kernel?

**Comment by g2gps** [comment]
@acassis I'll be happy to move the documentation over, once this PR is getting close to completion.

**Comment by g2gps** [comment]
In regards to the elf address locations. I still have two which I have stubbed out, like this:
```
diff --git a/libs/libc/machine/risc-v/arch_elf.c b/libs/libc/machine/risc-v/arch_elf.c
index 9efa483722..cc5c5579c3 100644
--- a/libs/libc/machine/risc-v/arch_elf.c
+++ b/libs/libc/machine/risc-v/arch_elf.c
@@ -413,7 +413,7 @@ int up_relocateadd(const Elf_Rela *rel, const Elf_Sym *sym,
           offset = (long)sym->st_value + (long)rel->r_addend - (long)addr;
           uint32_t val = _get_val((uint16_t *)addr) & 0xfffff000;
 
-          ASSERT(offset && val);
+          // ASSERT(offset && val);
 
           /* NOTE: we assume that a compiler adds an immediate value */
 
@@ -584,6 +584,10 @@ int up_relocateadd(const Elf_Rela *rel, const Elf_Sym *sym,
           *(uint16_t *)addr = (uint16_t)(sym->st_value + rel->r_addend);
         }
         break;
+      case R_RISCV_GOT_HI20:
+        {
+        }
+        break;
       default:
         berr("ERROR: Unsupported relocation: %ld\n",
              ARCH_ELF_RELTYPE(rel->r_info));
```
This currently 'works', at least I can load and run the applications I'm building. Although, it's not an ideal solution. Previously I've been using this [documentation](https://github.com/riscv-non-isa/riscv-elf-psabi-doc), but I'm unsure of the correct way to implement `R_RISCV_GOT_HI20`. Perhaps someone could provide some guidance?

**Comment by g2gps** [comment]
> This currently 'works', at least I can load and run the applications I'm building. Although, it's not an ideal solution. Previously I've been using this documentation, but I'm unsure of the correct way to implement R_RISCV_GOT_HI20. Perhaps someone could provide some guidance?

I've resolved this issue by building our toolchain with newlib. The relocation involving the GOT table were resulting from musl. See  #8886.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**